test/service/test_users.py

```diff
@@ -0,0 +1,205 @@
+# Copyright 2016, RadiantBlue Technologies, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+# this file except in compliance with the License. You may obtain a copy of the
+# License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed
+# under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+# CONDITIONS OF ANY KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations under the License.
+
+import logging
+import unittest
+import uuid
+from unittest.mock import patch, MagicMock, Mock
+from datetime import datetime
+
+
+import requests
+import requests_mock as rm
+from test import helpers
+from bfapi import piazza
+from bfapi.service import users
+from bfapi.db import DatabaseError
+from bfapi.config import GEOAXIS
+
+
+class AuthenticateViaGeoaxisTest(unittest.TestCase):
+    def setUp(self):
+        self._logger = logging.getLogger('bfapi.service.users')
+        self._logger.disabled = True
+        self._mockdb = helpers.mock_database()
+        self._logger = logging.getLogger('bfapi.service.jobs')
+        self._logger.disabled = True
+
+        self.mock_requests = rm.Mocker()  # type: rm.Mocker
+        self.mock_requests.start()
+        self.addCleanup(self.mock_requests.stop)
+        
+        self.userprofile_addr = 'https://{}/ms_oauth/resources/userprofile/me'.format(GEOAXIS)
+
+    def tearDown(self):
+        self._mockdb.destroy()
+        self._logger.disabled = False
+        
+    def create_mock(self, target_name):
+        patcher = patch(target_name)
+        self.addCleanup(patcher.stop)
+        return patcher.start()
+        
+    def test_throws_on_connection_error(self):
+        self.mock_request_get = self.create_mock('requests.get')
+        self.mock_request_get.side_effect = requests.ConnectionError()
+        with self.assertRaises(users.GeoaxisUnreachable):
+            users.authenticate_via_geoaxis('test-token')
+
+    def test_throws_on_unauthorized(self):
+        self.mock_requests.get(self.userprofile_addr, text='resp', status_code=401)
+        with self.assertRaises(users.GeoaxisUnauthorized):
+            users.authenticate_via_geoaxis('test-token')
+
+    def test_throws_on_http_error(self):
+        self.mock_requests.get(self.userprofile_addr, text='resp', status_code=405)
+        with self.assertRaises(users.GeoaxisError):
+            users.authenticate_via_geoaxis('test-token')
+        
+    def test_throws_on_no_uid(self):
+        self.mock_requests.get(self.userprofile_addr, text='{"username":"test-name"}', status_code=201)
```

Comment:
To the best of my knowledge it is.  It doesn't make `response.raise_for_status()` throw an error, whereas `status_code=401` and `status_cod=405` do.  I spent a while banging at this stuff to get it to the point where it seemed to work right, and I *think* I'm doing it the correct way.